fe/fe-core/src/main/java/com/starrocks/journal/JournalWriter.java

```diff
@@ -0,0 +1,121 @@
+    public static final Logger LOG = LogManager.getLogger(JournalWriter.class);
+    // other threads can put log to this queue by calling Editlog.logEdit()
+    private BlockingQueue<JournalQueueEntity> logQueue;
+    private Journal journal;
+
+    // store stagging log before commit
+    protected List<JournalQueueEntity> staggingLogs = new ArrayList<>();
+    // used for checking if edit log need to roll
+    protected long rollEditCounter = 0;
+
+    public JournalWriter(Journal journal, BlockingQueue<JournalQueueEntity> logQueue) {
+        super("JournalWriter", 0L);
+        this.journal = journal;
+        this.logQueue = logQueue;
+    }
+
+    @Override
+    protected void runOneCycle() {
+        try {
+            // 1. waiting if necessary until an element becomes available
+            JournalQueueEntity entity = logQueue.take();
+
+            long start = System.currentTimeMillis();
+            long uncommitedEstimatedBytes = 0;
+            staggingLogs.clear();
+
+            // 2. batch write
+            while (true) {
+                if (!staggingLogs.isEmpty()) {
+                    entity = logQueue.take();
+                }
+                // 2.1 write(no commit)
+                journal.writeWithinTxn(entity.getOp(), entity.getBuffer());
+                staggingLogs.add(entity);
+
+                // 2.2 check if is an emergency log
+                if (entity.getBetterCommitBeforeTime() > 0) {
+                    long delayMillis = System.currentTimeMillis() - entity.getBetterCommitBeforeTime();
+                    if (delayMillis >= 0) {
+                        LOG.info("log expect commit before {} is delayed {} mills, will commit now",
+                                entity.getBetterCommitBeforeTime(), delayMillis);
+                        break;
+                    }
+                }
+
+                // 2.3 check uncommitted logs by count
+                if (staggingLogs.size() >= Config.batch_journal_cnt) {
+                    LOG.info("staggging log {} >= {}, will commit now",
+                            staggingLogs.size(), Config.batch_journal_cnt);
+                    break;
+                }
+
+                // 2.4 check uncommitted logs by size
+                uncommitedEstimatedBytes += entity.estimatedSizeByte();
+                if (uncommitedEstimatedBytes >= Config.batch_journal_size_mb * 1024 * 1024) {
+                    LOG.info("uncommited estimated bytes {} >= {}MB, will commit now",
+                            uncommitedEstimatedBytes, Config.batch_journal_size_mb);
+                    break;
+                }
+
+                // 2.5 no more log in queue
+                if (logQueue.peek() == null) {
+                    break;
+                }
+            }
+
+            // 3. commit
+            journal.commitTxn();
+
+            // 4. countdown
+            for (JournalQueueEntity e : staggingLogs) {
+                e.countDown();
```

Comment:
how about returning a Future
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html

Reply:
Are you suggesting me to use another thread pool to countdown? I think this operation won't take much time. It's just a conditional variable.